src/script/microbit-interfacing/LogDataProcessor.ts

```diff
@@ -0,0 +1,204 @@
+    const view = new DataView(new ArrayBuffer(3));
+    // job
+    view.setUint8(0, this.job);
+    // request type; 0 = requestTypeNone, 1 = requestTypeLogLength, 2 = requestTypeLogRead
+    view.setUint8(1, this.requestType);
+    // format: 0 = HTML header; 1 = HTML; 2 = CSV.
+    view.setUint8(2, this.format);
+    return view;
+  };
+
+  private createLogReadDataView = (): DataView => {
+    if (!this.length) {
+      throw new Error('Missing log length. Cannot retrieve log data.');
+    }
+    this.batchReceived = 0;
+    this.batchLength = this.length - this.received;
+    if (this.batchLength > 19 * this.batch) {
+      this.batchLength = 19 * this.batch;
+    }
+    this.requestType = RequestType.RequestTypeLogRead;
+    // Force overflow to zero if needed.
+    if (this.job === 0xf0) {
+      this.job = 0;
+    } else {
+      this.job += 0x10;
+    }
+    this.jobLow = 0;
+
+    // https://github.com/microbit-foundation/microbit-ios/blob/2da377987e68e12983cb83c8c1989aa1884bf4b9/Source/MicroBitUtilityTypes.h#L30
+    const view = new DataView(new ArrayBuffer(16));
+    // job
+    view.setUint8(0, this.job);
+    // request type; 0 = requestTypeNone, 1 = requestTypeLogLength, 2 = requestTypeLogRead
+    view.setUint8(1, this.requestType);
+    // format: 0 = HTML header; 1 = HTML; 2 = CSV.
+    view.setUint8(2, this.format);
+    // reserved: set to zero.
+    view.setUint8(3, this.reserved);
+    // index: unsigned integer index into file.
+    view.setUint32(4, this.index, true);
+    // batchlen: unsigned size in bytes to return.
+    view.setUint32(8, this.batchLength, true);
+    // length: length of whole file, from request type 1 (Log file length)
+    view.setUint32(12, this.length, true);
+    return view;
+  };
+
+  private logLengthProcess(data: ArrayBuffer, dataLength: number): void {
+    if (dataLength < 4) {
+      throw new Error('No data in response');
+    }
+    // Is there a better way to do this?
+    const logLength = new DataView(data).getUint32(0, true);
+    this.length = logLength;
+    this.index = 0;
+    this.received = 0;
+
+    const view = this.createLogReadDataView();
+    this.requestWrite(view);
+  }
+
+  private logReadProcess(data: ArrayBuffer, dataLength: number): void {
+    if (dataLength < 1) {
+      throw new Error('No data in response');
+    }
+
+    // There's probably a better way to manage this?
+    if (this.data) {
+      const dataToAppend = new Uint8Array(data);
+      this.data = new Uint8Array([...this.data, ...dataToAppend]);
+    } else {
+      this.data = new Uint8Array(data);
+    }
+
+    this.received += dataLength;
+    this.batchReceived += dataLength;
+
+    if (this.batchReceived === this.batchLength) {
+      this.index += this.batchReceived;
+
+      if (this.received === this.length) {
+        const dataAsCsv = new TextDecoder().decode(this.data);
+        console.log('Imported log data:');
+        console.log(dataAsCsv);
```

Comment:
Just log out the data for now, but we'll need to validate it and ask the user whether to add or replace existing action data.

Reply:
There shouldn't be a problem making a new function in Gestures.ts or mlStore.ts to which we can pass this data. Or return data to the caller to handle there. We'll need to show a dialog.